File: src/main/java/oscar/cp/constraints/Sum.java
/*******************************************************************************
 * This file is part of OscaR (Scala in OR).
 *  
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 * 
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/gpl-3.0.html
 ******************************************************************************/

package oscar.cp.constraints;

import oscar.cp.core.CPOutcome;
import oscar.cp.core.CPPropagStrength;
import oscar.cp.core.CPVarInt;
import oscar.cp.core.Constraint;

/**
 * Sum Constraint: x[0]+x[1]+...+x[n] = y
 * @author Pierre Schaus pschaus@gmail.com
 */
public class Sum extends Constraint {
	
	private CPVarInt [] x;
	private CPVarInt y;

	public Sum(CPVarInt [] x, CPVarInt y) {
		super(x[0].getStore(),"Sum");
		this.x = x;
		this.y = y;
		setIdempotent();
	}

    /**
     * y == sum(x)
     * @param x
     * @param y
     */
	public Sum(CPVarInt [] x, int y) {
		this(x,new CPVarInt(x[0].getStore(),y,y));
	}

	@Override
	protected CPOutcome setup(CPPropagStrength l) {
		if (propagate() == CPOutcome.Failure) {
			return CPOutcome.Failure;
		}
		for (int i = 0; i < x.length; i++) {
			if (!x[i].isBound()) 
				x[i].callPropagateWhenBoundsChange(this);
		}
		if (!y.isBound())
			y.callPropagateWhenBoundsChange(this);
		return CPOutcome.Suspend;
	}
	
	@Override
	protected CPOutcome propagate() {
		int maxsumx = 0;
		int minsumx = 0;
		for (int i = 0; i < x.length; i++) {
			maxsumx += x[i].getMax();
			minsumx += x[i].getMin();
		}
		
		if (y.updateMax(maxsumx) == CPOutcome.Failure) {
			return CPOutcome.Failure;
		}
		
		if (y.updateMin(minsumx) == CPOutcome.Failure) {
			return CPOutcome.Failure;
		}
		
		for (int i = 0; i < x.length; i++) {
			int maxsumxi = maxsumx - x[i].getMax();
			int minsumxi = minsumx - x[i].getMin();
			int maxi = y.getMax() - minsumxi;
			int mini = y.getMin() - maxsumxi;
			if (x[i].updateMax(maxi) == CPOutcome.Failure) {
				return CPOutcome.Failure;
			}
			if (x[i].updateMin(mini) ==  CPOutcome.Failure) {
				return CPOutcome.Failure;
			}
		}
			
		return CPOutcome.Suspend;
	}
	
	

}

File: oscar-cp/src/main/scala/oscar/cp/constraints/Sum.scala
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.constraints

import oscar.cp.core._
import oscar.algo.reversible._
import oscar.cp.core.CPOutcome._
import math.max

/**
 * Implementation of Sum Constraint:
 * @author Pierre Schaus pschaus@gmail.com
 * @author Steven Gay steven.gay@uclouvain.be
 */

class Sum(val X: Array[CPIntVar], val y: CPIntVar) extends Constraint(y.store, "Sum") {

  val x = X.map(i => i)
  val xSize = x.size
  val sumBoundVars = new ReversibleInt(s, 0)
  val nBoundVars = new ReversibleInt(s, 0)
  
  idempotent = true

  override def setup(l: CPPropagStrength): CPOutcome = {
    
    priorityL2 = CPStore.MAXPRIORL2 - 1
    X.foreach(_.callPropagateWhenBoundsChange(this, false))
    y.callPropagateWhenBoundsChange(this, false)
    val oc = propagate()
    //println("oc:"+oc)
    oc
  }

  private def setBound(i: Int) {
    sumBoundVars.value = sumBoundVars.value + x(i).value
    val tmp = x(nBoundVars.value)
    x(nBoundVars.value) = x(i)
    x(i) = tmp
    nBoundVars.incr()
  }

  override def propagate(): CPOutcome = {
    // invariant to maintain: sumBoundVars is the sum of variables x(0 until nBounds), which have to be bound.

    var sumxmin = 0 
    var sumxmax = 0
    var maxDiff = 0
    var reduce = true
    var i = 0
    
    while(reduce) {
      reduce = false
      sumxmin = sumBoundVars.value
      sumxmax = sumBoundVars.value

      // step 1: filter bound variables and get maximum range size 
      var i = nBoundVars.value
      while (i < xSize) {
        val ximax = x(i).max
        val ximin = x(i).min
        
        sumxmin += ximin
        sumxmax += ximax
        maxDiff = max(ximax - ximin, maxDiff)
        
        if (ximax == ximin) setBound(i)
        i += 1
      }
      
      
      // step 2: propagate from x to y
      if (y.updateMax(sumxmax) == Failure) return Failure
      if (y.updateMin(sumxmin) == Failure) return Failure

      
      // step 3: propagate from y to x: if there is a ximax that makes the sum go over y, find and trim it
      val ymax = y.max
      if (sumxmin + maxDiff > ymax) {
        i = nBoundVars.value
        while (i < xSize) {
          // Compute new upper bound for xi
          val oldximax = x(i).max
          val ymini = sumxmin - x(i).min
          val ximax = ymax - ymini

          // Update and check whether something changed
          if (ximax < oldximax) {
            if (x(i).updateMax(ximax) == Failure) return Failure
            
            val newximax = x(i).max
            val xidiff = newximax - oldximax
            sumxmax += xidiff
            if (newximax < ximax) reduce = true   // this can happen when x(i) is a multiplicative or non-bijective view ; then do fixpoint here.
          }
          i += 1
        }
      }
      
      // step 4: same for ximin
      val ymin = y.min
      if (sumxmax - maxDiff < ymin) {
        i = nBoundVars.value
        while (i < xSize) {
          val oldximin = x(i).min
          val ymaxi = sumxmax - x(i).max
          val ximin = ymin - ymaxi
          
          if (ximin > oldximin) {
            if (x(i).updateMin(ximin) == Failure) return Failure
            
            val newxmin = x(i).min
//            val xidiff = newxmin - oldximin
//            sumxmin += xidiff
            if (newxmin > ximin) reduce = true
          }
          i += 1
        }
      }
      
  }
    
  Suspend
  }

}


//          val oc = x(i).updateMax(ximax) // cannot fail
//          assert(oc != Failure)

          // val oc = x(i).updateMin(ximin) // cannot fail
          // assert(oc != Failure)





















